File: core/src/main/java/org/vizzini/core/TimePrinter.java
package org.vizzini.core;

import org.apache.commons.lang3.StringUtils;

/**
 * Provides convenience methods to print time.
 */
public final class TimePrinter
{
    /** Seconds to milliseconds factor. */
    private final int SECONDS_TO_MS = 1000;

    /** Minutes to milliseconds factor. */
    private final int MINUTES_TO_MS = 60 * SECONDS_TO_MS;

    /**
     * @param title Title.
     * @param start Start time. (ms)
     * @param end End time. (ms)
     * 
     * @return a formatted string.
     */
    public String formatElapsedTime(final String title, final long start, final long end)
    {
        final long myStart = Math.min(start, end);
        final long myEnd = Math.max(start, end);

        final long elapsed = myEnd - myStart;
        final int minutes = (int)(elapsed / MINUTES_TO_MS);
        final long leftover = elapsed - (minutes * MINUTES_TO_MS);
        final int seconds = (int)(leftover / SECONDS_TO_MS);

        final StringBuilder sb = new StringBuilder();

        sb.append(createTitleString(title));
        sb.append(minutes);
        sb.append(":");
        sb.append(StringUtils.leftPad(String.valueOf(seconds), 2, "0"));
        sb.append(" (").append(elapsed).append(" ms)");

        return sb.toString();
    }

    /**
     * @param title Title.
     * @param start Start time. (ms)
     * @param end End time. (ms)
     */
    public void printElapsedTime(final String title, final long start, final long end)
    {
        System.out.println(formatElapsedTime(title, start, end));
    }

    /**
     * @param title Title.
     * 
     * @return a new title string.
     */
    private String createTitleString(final String title)
    {
        final StringBuilder sb = new StringBuilder();

        if (StringUtils.isEmpty(title))
        {
            sb.append("Elapsed time ");
        }
        else
        {
            sb.append(title);
            sb.append(" elapsed time ");
        }

        return sb.toString();
    }
}

File: coreweb/src/main/js/util/TimePrinter.js
/*
 * Provides convenience methods to print time.
 */
var TimePrinter =
{
    /* Seconds to milliseconds factor. */
    SECONDS_TO_MS: 1000,

    /* Minutes to milliseconds factor. */
    MINUTES_TO_MS: 60 * 1000,

    /*
     * @param title Title. 
     * @param start Start time. (ms) 
     * @param end End time. (ms)
     * 
     * @return a formatted string.
     */
    formatElapsedTime: function(title, start, end)
    {
        var myStart = Math.min(start, end);
        var myEnd = Math.max(start, end);

        var elapsed = myEnd - myStart;
        var minutes = Math.floor(elapsed / this.MINUTES_TO_MS);
        var leftover = elapsed - (minutes * this.MINUTES_TO_MS);
        var seconds = Math.floor(leftover / this.SECONDS_TO_MS);

        var sb = "";

        sb += this.createTitleString(title);
        sb += minutes;
        sb += ":";

        if (seconds < 10)
        {
            sb += "0";
        }

        sb += seconds;
        sb += " (";
        sb += elapsed;
        sb += " ms)";

        return sb;
    },

    /*
     * @param title Title.
     * @param start Start time. (ms)
     * @param end End time. (ms)
     */
    printElapsedTime: function(title, start, end)
    {
        console.info(this.formatElapsedTime(title, start, end));
    },

    /*
     * @param title Title.
     * 
     * @return a new title string.
     */
    createTitleString: function(title)
    {
        var sb = "";

        if (!title || title.length === 0)
        {
            sb += "Elapsed time ";
        }
        else
        {
            sb += title;
            sb += " elapsed time ";
        }

        return sb;
    },
}
